jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java

```diff
@@ -110,32 +114,53 @@ static void transitiveReduction(BitSet[] pathMatrix)
     }
 
     /**
-     * This method will remove all transitive edges from the graph passed as input parameter.
-     *
+     * This method will remove all transitive edges from the DAG (directed acyclic graph) graph passed as input
+     * parameter. If you need to transitively reduce a cyclic graph, please use {@link CyclicTransitiveReduction}
+     * instead.
      * <p>
      * You may want to clone the graph before, as transitive edges will be pitilessly removed.
      * </p>
-     *
-     * e.g.
-     *
-     * <pre>
-     * {
-     *     &#64;code DirectedGraph&lt;V, T&gt; soonToBePrunedDirectedGraph;
-     *
-     *     TransitiveReduction.INSTANCE.reduce(soonToBePrunedDirectedGraph);
-     *
-     *     // pruned !
-     * }
+     * e.g. <pre>
+     * DirectedGraph&lt;V, T&gt; soonToBePrunedDirectedGraph;
+     * TransitiveReduction.INSTANCE.reduce(soonToBePrunedDirectedGraph);
      * </pre>
      *
      * @param directedGraph the directed graph that will be reduced transitively
      * @param <V> the graph vertex type
      * @param <E> the graph edge type
+     * @throws IllegalArgumentException if the given directed graph contains cycles, because this algorithm only works
+     *                                  for acyclic graphs
      */
     public <V, E> void reduce(final Graph<V, E> directedGraph)
+    {
+        reduce(directedGraph, true);
+    }
+
+    /**
+     * This method is just a special version of {@link #reduce(Graph)} with the additional option to avoid checking the
+     * input graph for cycles. It is package-scoped in order to allow other classes in the same package, namely
```

Comment:
For now the method is package-scoped because it only is used internally. I applied the need-to-know principle and made it non-public for now because I see no obvious reason for normal users other than contributors to use it. Until now there was no check for that case (a pretty cheap one, at that) but a huge bug. I would recommend to avoid premature optimisation similar to avoiding `synchronized` in favour of documenting that methods should not be used concurrently. Instead I recommend for such a big and widely used tools to be programmed more defensively and have the end user in mind first and foremost.

Reply:
Thanks for your opinion, but I think the benefit of exposing the parameter is clear, and the default value makes it safe.

Reply:
Feel free to change access modifiers to your heart's content after you will have accepted the PR.

Reply:
No worries, I'll leave that for the mythical "maintainers".

Reply:
I was not referring to anyone mythical or making up things, see https://github.com/jgrapht/jgrapht/pull/1008#discussion_r535747879 for an explanation of the term "maintainer".